File: ueueo-framework/ueueo-threading/src/main/java/Abp/Threading/AmbientDataContextAmbientScopeProvider.cs
using System;
using System.Collections.Concurrent;
using System.Collections.Generic;
using JetBrains.Annotations;
using Microsoft.Extensions.Logging;
using Microsoft.Extensions.Logging.Abstractions;

namespace Volo.Abp.Threading;

public class AmbientDataContextAmbientScopeProvider<T> : IAmbientScopeProvider<T>
{
    public ILogger<AmbientDataContextAmbientScopeProvider<T>> Logger;// { get; set; }

    private static readonly ConcurrentDictionary<String, ScopeItem> ScopeDictionary = new ConcurrentDictionary<String, ScopeItem>();

    private readonly IAmbientDataContext _dataContext;

    public AmbientDataContextAmbientScopeProvider(@NonNull IAmbientDataContext dataContext)
    {
        Objects.requireNonNull(dataContext, nameof(dataContext));

        _dataContext = dataContext;

        Logger = NullLogger<AmbientDataContextAmbientScopeProvider<T>>.Instance;
    }

    public T GetValue(String contextKey)
    {
        var item = GetCurrentItem(contextKey);
        if (item == null)
        {
            return default;
        }

        return item.Value;
    }

    public IDisposable BeginScope(String contextKey, T value)
    {
        var item = new ScopeItem(value, GetCurrentItem(contextKey));

        if (!ScopeDictionary.TryAdd(item.Id, item))
        {
            throw new AbpException("Can not add item! ScopeDictionary.TryAdd returns false!");
        }

        _dataContext.SetData(contextKey, item.Id);

        return new DisposeAction(() =>
        {
            ScopeDictionary.TryRemove(item.Id, out item);

            if (item.Outer == null)
            {
                _dataContext.SetData(contextKey, null);
                return;
            }

            _dataContext.SetData(contextKey, item.Outer.Id);
        });
    }

    private ScopeItem GetCurrentItem(String contextKey)
    {
        return _dataContext.GetData(contextKey) is String objKey ? ScopeDictionary.GetOrDefault(objKey) : null;
    }

    private class ScopeItem
    {
        public String Id;//  { get; }

        public ScopeItem Outer;//  { get; }

        public T Value;//  { get; }

        public ScopeItem(T value, ScopeItem outer = null)
        {
            Id = Guid.NewGuid().ToString();

            Value = value;
            Outer = outer;
        }
    }
}

File: ueueo-framework/ueueo-threading/src/main/java/com/ueueo/threading/AmbientDataContextAmbientScopeProvider.java
package com.ueueo.threading;

import com.ueueo.DisposeAction;
import com.ueueo.IDisposable;
import lombok.Getter;
import lombok.extern.slf4j.Slf4j;
import org.springframework.lang.NonNull;

import java.util.Objects;
import java.util.UUID;
import java.util.concurrent.ConcurrentHashMap;

@Slf4j
public class AmbientDataContextAmbientScopeProvider<T> implements IAmbientScopeProvider<T> {

    private static ConcurrentHashMap<String, ScopeItem> ScopeDictionary = new ConcurrentHashMap<>();

    private IAmbientDataContext dataContext;

    public AmbientDataContextAmbientScopeProvider(@NonNull IAmbientDataContext dataContext) {
        Objects.requireNonNull(dataContext);
        this.dataContext = dataContext;
    }

    @Override
    public T getValue(String contextKey) {
        ScopeItem item = getCurrentItem(contextKey);
        if (item == null) {
            return null;
        }
        return (T) item.value;
    }

    @Override
    public IDisposable beginScope(String contextKey, T value) {
        ScopeItem item = new ScopeItem(value, getCurrentItem(contextKey));
        ScopeDictionary.put(item.id, item);
        dataContext.setData(contextKey, item.id);
        return new DisposeAction(() -> {
            ScopeItem i = ScopeDictionary.remove(item.getId());
            if (i.getOuter() == null) {
                dataContext.setData(contextKey, null);
                return;
            }
            dataContext.setData(contextKey, i.getOuter().id);
        });
    }

    private ScopeItem getCurrentItem(String contextKey) {
        if (dataContext.getData(contextKey) instanceof String) {
            String objKey = (String) dataContext.getData(contextKey);
            return ScopeDictionary.get(objKey);
        }
        return null;
    }

    @Getter
    private static class ScopeItem {
        private final String id;
        private final ScopeItem outer;
        private final Object value;

        public ScopeItem(Object value, ScopeItem outer) {
            this.id = UUID.randomUUID().toString();
            this.value = value;
            this.outer = outer;
        }
    }
}
